Extensions/OpenIdConnectMiddlewareDiagnostics.cs

```diff
@@ -0,0 +1,184 @@
+        //
+        // Summary:
+        //     Invoked if exceptions are thrown during request processing. The exceptions will
+        //     be re-thrown after this event unless suppressed.
+        static Func<AuthenticationFailedContext, Task> onAuthenticationFailed;
+        //
+        // Summary:
+        //     Invoked when a request is received on the RemoteSignOutPath.
+        static Func<RemoteSignOutContext, Task> onRemoteSignOut;
+        //
+        // Summary:
+        //     Invoked before redirecting to the identity provider to sign out.
+        static Func<RedirectContext, Task> onRedirectToIdentityProviderForSignOut;
+        //
+        // Summary:
+        //     Invoked before redirecting to the Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectOptions.SignedOutRedirectUri
+        //     at the end of a remote sign-out flow.
+        static Func<RemoteSignOutContext, Task> onSignedOutCallbackRedirect;
+ 
+
+        /// <summary>
+        /// Subscribes to all the OpenIdConnect events, to help debugging, while
+        /// preserving the previous handlers (which are called)
+        /// </summary>
+        /// <param name="events">Events to subscribe to</param>
+        public static void Subscribe(OpenIdConnectEvents events)
+        {
+            onRedirectToIdentityProvider = events.OnRedirectToIdentityProvider;
+            events.OnRedirectToIdentityProvider = OnRedirectToIdentityProvider;
+
+            onMessageReceived = events.OnMessageReceived;
+            events.OnMessageReceived = OnMessageReceived;
+
+            onAuthorizationCodeReceived = events.OnAuthorizationCodeReceived;
+            events.OnAuthorizationCodeReceived = OnAuthorizationCodeReceived;
+
+            onTokenResponseReceived = events.OnTokenResponseReceived;
+            events.OnTokenResponseReceived = OnTokenResponseReceived;
+
+            onTokenValidated = events.OnTokenValidated;
+            events.OnTokenValidated = OnTokenValidated;
+
+            onUserInformationReceived = events.OnUserInformationReceived;
+            events.OnUserInformationReceived = OnUserInformationReceived;
+
+            onAuthenticationFailed = events.OnAuthenticationFailed;
+            events.OnAuthenticationFailed = OnAuthenticationFailed;
+
+            onRemoteSignOut = events.OnRemoteSignOut;
+            events.OnRemoteSignOut = OnRemoteSignOut;
+
+            onRedirectToIdentityProviderForSignOut = events.OnRedirectToIdentityProviderForSignOut;
+            events.OnRedirectToIdentityProviderForSignOut = OnRedirectToIdentityProviderForSignOut;
+
+            onSignedOutCallbackRedirect = events.OnSignedOutCallbackRedirect;
+            events.OnSignedOutCallbackRedirect = OnSignedOutCallbackRedirect;
+        }
+
+        static async Task OnRedirectToIdentityProvider(RedirectContext context)
+        {
+            Debug.WriteLine($"1. Begin {nameof(OnRedirectToIdentityProvider)}");
```

Comment:
I think you could tap into the asp.net core logging https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-2.2

Reply:
yes, @bgavrilMS this would be better.